google/cloud/bigtable/examples/stock_backtesting/backtesting_main.cc

```diff
@@ -0,0 +1,258 @@
+                                        absl::CivilDay const end_date) {
+  std::vector<std::string> row_keys;
+
+  // In Bigtable each row is per {ticker + year}.
+  auto const start_year = start_date.year();
+  auto const end_year = end_date.year();
+  for (auto i = start_year; i <= end_year; ++i) {
+    row_keys.emplace_back(absl::StrCat(ticker, kRowKeyDelimiter, i));
+  }
+
+  return row_keys;
+}
+
+cbt::Filter PrepareFilter(cbt::examples::Strategy const& strategy,
+                          absl::CivilDay const start_date,
+                          absl::CivilDay const end_date) {
+  // For one specific row key, we want to query the |base| and |sample| columns
+  // specified in strategy, and the timestamp of [start_date, end_date].
+  // Use filter combo to implement this.
+
+  // Column filters, only select the targeting columns.
+  std::vector<cbt::Filter> column_filters;
+  std::set<std::string> existing_signal;
+  for (auto const& condition : strategy.conditions()) {
+    if (existing_signal.find(condition.base()) == existing_signal.end()) {
+      existing_signal.insert(condition.base());
+      std::vector<std::string> split =
+          absl::StrSplit(condition.base(), kColumnDelimiter);
+      // In the format of column_family::column_qualifier.
+      if (split.size() != 2) {
+        throw std::runtime_error("Invalid strategy definition.");
+      }
+      column_filters.emplace_back(
+          cbt::Filter::ColumnName(split.front(), split.back()));
+    }
+
+    if (existing_signal.find(condition.sample()) == existing_signal.end()) {
+      existing_signal.insert(condition.sample());
+      std::vector<std::string> split =
+          absl::StrSplit(condition.sample(), kColumnDelimiter);
+      // In the format of column_family::column_qualifier.
+      if (split.size() != 2) {
+        throw std::runtime_error("Invalid strategy definition.");
+      }
+      column_filters.emplace_back(
+          cbt::Filter::ColumnName(split.front(), split.back()));
+    }
+  }
+
+  cbt::Filter column_filter = cbt::Filter::InterleaveFromRange(
+      column_filters.begin(), column_filters.end());
+
+  // Timestamp filter, only select cells in the target window.
+  auto const start_time = absl::FromCivil(start_date, absl::UTCTimeZone());
+  auto const end_time = absl::FromCivil(end_date, absl::UTCTimeZone());
+  auto const start_milli =
+      absl::ToChronoMilliseconds(start_time - absl::UnixEpoch());
+  auto const end_milli =
+      absl::ToChronoMilliseconds(end_time - absl::UnixEpoch());
+  cbt::Filter timestamp_filter =
+      cbt::Filter::TimestampRange(start_milli, end_milli);
+
+  return cbt::Filter::Chain(column_filter, timestamp_filter);
+}
+
+void CalculateProfit(
+    std::unordered_map<std::string, std::map<absl::CivilDay, double>> const&
+        time_series,
+    cbt::examples::Strategy const& strategy) {
+  // Assume the base and sample fields are the same across conditions.
```

Comment:
Strange....

Reply:
(Commenting, so I don't forget)

We wouldn't need to make this assumption if our proto looked like this: 
```proto
message Condition {
  double threshold = 1;
  double money = 2;
}

message Strategy {
  string base = 1;
  string sample = 2;
  
  repeated Condition conditions = 3;
}
```